src/jit.jl

```diff
@@ -194,13 +194,38 @@ function add_entry!(mod::LLVM.Module, func::ANY, tt::ANY; kernel::Bool=false)
                     # this only happens when codegen wants to pass a pointer.
                     @assert isa(codegen_t, LLVM.PointerType)
                     @assert eltype(codegen_t) == wrapper_t
+
                     # copy the argument value to a stack slot, and reference it.
                     ptr = alloca!(builder, wrapper_t)
-                    store!(builder, wrapper_param, ptr)
                     if LLVM.addrspace(codegen_t) != 0
                         ptr = addrspacecast!(builder, ptr, codegen_t)
                     end
+                    store!(builder, wrapper_param, ptr)
                     push!(wrapper_args, ptr)
+
+                    # Julia marks parameters as TBAA immutable;
+                    # this is incompatible with us storing to a stack slot, so clear TBAA
+                    # TODO: tag with alternative information (eg. TBAA, or invariant groups)
+                    entry_params = collect(parameters(entry_f))
+                    candidate_uses = []
+                    for param in entry_params
+                        append!(candidate_uses, collect(uses(param)))
+                    end
+                    while !isempty(candidate_uses)
+                        usepair = shift!(candidate_uses)
+                        inst = user(usepair)
+
+                        if hasmetadata(inst, LLVM.MD_tbaa)
+                            metadata!(inst, LLVM.MD_tbaa)
```

Comment:
This seems odd to me. The purpose of this according to the comment above is to clear the TBAA metadata, right?
Does `metadata!` clear the associated metadata if supplied twice?

Reply:
Ah `metadata!(inst, kind)` sets the associated metadata to `C_NULL` so it does clear it. Might be worth a comment or documentation in LLVM.jl?

Reply:
Yeah, I wasn't sure about a proper interface. Maybe I should wrap the metadata in a dict-like structure where `empty!` is more explicit?

Reply:
Yeah, there's a bunch of similar artificial structs in LLVM already, this is a good place for another. I'll have a look tomorrow.

Reply:
https://github.com/maleadt/LLVM.jl/pull/49/commits/224a2a8b4135c07a33a22f710bada3ff0582c75f